### PR TITLE
3.3.5 paymaster

### PR DESCRIPTION
**Pull request by danjohnso**
Not sure how to PR for a new branch/custom feature, but I will start here and can change as needed.

**Changes proposed**:

- Add paymaster AI for Paymaster creature
- Options include:
  - Add gold, amount is configurable
  - Learn professions and all recipes associated with them
  - Add reputation for outland and northrend factions

**Target branch(es)**: 335

**Tests performed**: (Does it build, tested in-game, etc)

Built and added creature in world.

image

**Known issues and TODO list**: None

### CODE REVIEW

**Comment by Aokromes** [comment]
cool, it's like TBC classic ptr server.

**Comment by danjohnso** [comment]
> 

That was the inspiration, but the NPC has been in the creature template since vanilla, it just never has an implementation.  Added 3.3.5 notable things like inscription and northrend factions in addition to the TBC PTR functionality.  I am having my friends test it some more yet, may update tonight if they find issues

**Comment by danjohnso** [comment]
@Aokromes what do I need to change for the custom_script_loader to be happy?  Mine looks the same as other custom modules as far as I can tell

**Comment by Aokromes** [comment]
> @Aokromes what do I need to change for the custom_script_loader to be happy? Mine looks the same as other custom modules as far as I can tell

the current system is a pain to maintain with multiple patches, you need to set your changes between the // on unused ranks by other patches.

File: src/server/scripts/Custom/custom_script_loader.cpp [hunk 1]
```diff
@@ -19,7 +19,6 @@
 
 // The name of this function should match:
 // void Add${NameOfDirectory}Scripts()
-
```
Comment (Aokromes) [inline src/server/scripts/Custom/custom_script_loader.cpp]: revert this :)

File: sql/custom/world/world.Paymaster.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+UPDATE `creature_template` SET `ScriptName`="npc_paymaster" WHERE `Entry`=26075;
```
Comment (Aokromes) [inline sql/custom/world/world.Paymaster.sql]: missing newline.

Comment (danjohnso) [reply sql/custom/world/world.Paymaster.sql]: Done

File: src/server/scripts/Custom/custom_script_loader.cpp (hunk 1 above)
Comment (danjohnso) [reply src/server/scripts/Custom/custom_script_loader.cpp]: Done

File: src/server/worldserver/worldserver.conf.dist [hunk 3]
```diff
@@ -4147,3 +4147,26 @@ Metric.OverallStatusInterval = 1
...
+Paymaster.Announce.Enable = 1
+
+#    Paymaster.AddGold
+#         Configurable gold added amount from Paymaster
+#         Default:    5000
+
+Paymaster.AddGold = 5000
+
+#
+###################################################################################################
```
Comment (acidmanifesto) [inline src/server/worldserver/worldserver.conf.dist]: line break preferrably, but other then that, this is really great stuff. Great job.

**Review by acidmanifesto — APPROVED** [review]

File: src/server/worldserver/worldserver.conf.dist (hunk 3 above)
Comment (danjohnso) [reply src/server/worldserver/worldserver.conf.dist]: A line break after 4172?

Comment (acidmanifesto) [reply src/server/worldserver/worldserver.conf.dist]: yes
image
I believe it deals with the code style of the project. but it isnt anything project breaking.

Comment (danjohnso) [reply src/server/worldserver/worldserver.conf.dist]: Yup, updated and finally fixed all the other random whitespace that appeared...seems like the cmake generation for visual studio isn't copying the editor config to the right spot, might have to look at that

**Comment by danjohnso** [comment]
Any other changes you want @Aokromes ?

**Comment by danjohnso** [comment]
Anything else to change here @Aokromes ?

File: src/server/scripts/Custom/Paymaster.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+#include "Player.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "ScriptedGossip.h"
+#include "SharedDefines.h"
+#include "World.h"
+#include "WorldSession.h"
+#include "DataStores/DBCStores.h"
+#include "Spells/SpellMgr.h"
+#include <Commands/cs_learn.cpp>
```
Comment (jackpoz) [inline src/server/scripts/Custom/Paymaster.cpp]: I'm not really a fan of including cpp files. In this particular case just copy `HandleLearnSkillRecipesHelper` to this file with its implementation and remove the cpp include.

**Review by jackpoz — CHANGES_REQUESTED** [review]

File: src/server/scripts/Custom/Paymaster.cpp [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+#include "ObjectMgr.h"
+#include "Player.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "ScriptedGossip.h"
+#include "SharedDefines.h"
+#include "World.h"
+#include "WorldSession.h"
+#include "DataStores/DBCStores.h"
+#include "Spells/SpellMgr.h"
```
Comment (jackpoz) [inline src/server/scripts/Custom/Paymaster.cpp]: the folder is not needed
```suggestion
#include "SpellMgr.h"
```

File: src/server/scripts/Custom/Paymaster.cpp [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+#include "ObjectAccessor.h"
+#include "ObjectMgr.h"
+#include "Player.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "ScriptedGossip.h"
+#include "SharedDefines.h"
+#include "World.h"
+#include "WorldSession.h"
+#include "DataStores/DBCStores.h"
```
Comment (jackpoz) [inline src/server/scripts/Custom/Paymaster.cpp]: the folder is not needed

```suggestion
#include "DBCStores.h"
```

**Comment by jackpoz** [comment]
There are a couple of things to adjust, please test the PR again ingame after having dealt with the comments and then we can merge it

File: src/server/scripts/Custom/Paymaster.cpp (hunk 4 above)
Comment (danjohnso) [reply src/server/scripts/Custom/Paymaster.cpp]: Resolved this and the include references in latest commit.  Tested in game, everything still works great

**Review by acidmanifesto — APPROVED** [review]

**Review by jackpoz — APPROVED** [review]

**Comment by jackpoz** [comment]
I added this branch to https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki with a short description. Please let me know if I should change the description

**Pull request merged**